### ISSUE TITLE
Training with ncc loss leads to nan and inf losses

### ISSUE THREAD

**Issue by theoVag** [voxelmorph/voxelmorph#127]
Hello,
I am using the cvpr s2s with train.py file with mri images normalized to [0,1] and affine registered.
When i use mse as loss function the loss from the beginning is very low and  the loss is changing in the 3rd decimal. And eventually it is just oscillating around these values. The result in registration is just satisfying.
Then  tried to use ncc loss in order to check if I can improve registration but after 5-10 epochs the losses becomes inf and nan as you can see below:
![Screenshot from 2019-11-25 18-59-22](https://user-images.githubusercontent.com/32677031/69564128-15f68580-0fbb-11ea-9e9c-3712771ef4d5.png)

Do you have any ideas why is this happening?

**Comment by adalca** [comment]
Hi @theoVag 

From the screenshot, it seems like you are loading existing models, so it's not that surprising that the loss starts low and doesn't change much. Can you provide the exact parameters (command line) you ran for the second case? there's been one or two questions (#13, #82) that referred to nans, but if i recall correctly they were alleviate by correct parameter values.

**Comment by theoVag** [comment]
Hello,
I tried different values for reg param between 0.01-1000 but the difference between them is only that nan and inf appears in different epochs. I observed that with reg param=1000 the nan popped up in 57 epoch so when I increase reg_param he nan happens later. But by increasing reg param I take smaller values of flow and the network doesn't learn. I used model in 56 epoch to register and the result was not good.
I run the training wth the parameters below:

train (data_dir = ".../train/vols",model_dir =".../voxelmorph-master/models",load_model_file= None, model='vm2',reg_param=1000,steps_per_epoch=100,data_loss='ncc', gpu_id= '0', lr= 0.0001, nb_epochs= 80, batch_size= 1,initial_epoch=0)

My data are a part of Osteoarthritis Initiative with mri from knee.
I also provide you the csv with losses ( https://pastebin.com/r5vHkFiQ ). Is there anything that I can do to avoid this error and train with ncc?

**Comment by junyuchen245** [comment]
Hi @adalca @theoVag,

I encountered a similar problem with the ncc loss, I was getting nan after a few epoch. I believe that nothing is wrong with the ncc loss implementation since the epsilon (1e-5) was added to prevent the loss becomes nan. So, I thought if the problem is in the 'nrn_layers.SpatialTransformer', and I implemented my own 3D spatial transformer based on the 2D tf.contrib.image.dense_image_warp (http://tensorflow.biotecan.com/python/Python_1.8/tensorflow.google.cn/api_docs/python/tf/contrib/image/dense_image_warp.html). Now everything works great! I'm not sure what the problem is with the 'nrn_layers.SpatialTransformer'. Maybe you could look into it?

**Comment by adalca** [comment]
Hi @theoVag 
based on your other messages I assume you've gotten past this issue, but let us know if they are still happening.

**Comment by 907613871** [comment]
> Hello,
> I am using the cvpr s2s with train.py file with mri images normalized to [0,1] and affine registered.
> When i use mse as loss function the loss from the beginning is very low and the loss is changing in the 3rd decimal. And eventually it is just oscillating around these values. The result in registration is just satisfying.
> Then tried to use ncc loss in order to check if I can improve registration but after 5-10 epochs the losses becomes inf and nan as you can see below:
> ![Screenshot from 2019-11-25 18-59-22](https://user-images.githubusercontent.com/32677031/69564128-15f68580-0fbb-11ea-9e9c-3712771ef4d5.png)
> 
> Do you have any ideas why is this happening?

I also encountered the same problem. When using ‘ncc’ there will be -inf and nan, but I can’t learn anything with mse training. I wonder if you can introduce your solution. Thank you very much.

**Comment by adalca** [comment]
@907613871  Can you let us know which version of vxm you are using (legacy or new master branch), and what learning date you are using?

**Comment by 907613871** [comment]
> @ 907613871 您能告诉我们您使用的是哪个版本的vxm（旧版或新的master分支），以及您使用的是什么学习日期？

I am very happy to receive your reply, I am using CVPR2018.Use the latest version of the official website will also produce this situation. The data size is 96*112*96. Changing the learning rate multiple times will produce -inf. The network cannot learn anything when using mse. Thank you!

**Comment by adalca** [comment]
@907613871 

can you please try the `NCC` class from the `dev` branch: https://github.com/adalca/neurite/blob/dev/neurite/tf/losses.py
We've recently made some modifications that seemed to improve stability.

Thanks

**Comment by 907613871** [comment]
> @907613871
> 
> can you please try the `NCC` class from the `dev` branch: https://github.com/adalca/neurite/blob/dev/neurite/tf/losses.py
> We've recently made some modifications that seemed to improve stability.
> 
> Thanks
Why can't I learn anything with mse?

**Comment by adalca** [comment]
> Why can't I learn anything with mse?

I don't know, what does "can't learn anything" mean? what is your data? It owuld be good to have more details

**Comment by 907613871** [comment]
> > 为什么我不能通过MSE学习任何东西？
> 
> 我不知道，“什么都学不到”是什么意思？你的数据是什么？有更多详细信息应该很好

I use multiple brain MR data sets with a size of 96*112*96. Using MSE, there is no change in the warped image after multiple iterations.

**Comment by adalca** [comment]
@907613871 we need way more details :) For example
- what is the intensity range of the images?
- what weight are you using for the regularization?
- how many iterations (can we see a graph)

**Comment by 907613871** [comment]
> @ 907613871我们需要更多细节:)例如
> 
> * 图像的强度范围是多少？
> * 您使用什么权重进行正则化？
> * 多少次迭代（我们可以看到一个图）

Image intensity (1.0, 1.0, 1.0), regularization weight 0.01, learning rate 1e-4, 1e-5, 1e-6 have all tried. There are 100,000 iterations, but in fact there is no change from about 20,000 iterations.

**Comment by adalca** [comment]
@SLKaMiHi 

what do you mean by 
> Image intensity (1.0, 1.0, 1.0),
why three "1.0"?

overall, though that should work with MSE. Can you show me a plot of the curve and exmaple input/output images?
